### PR TITLE
feat: Least-Loaded Expert Parallelism with new upstream

### PR DESCRIPTION
**Pull request by xrsrke**
## Summary

Implements **Least-Loaded Expert Parallelism (LLEP)**, a dynamic load-balancing strategy for Expert Parallelism in MoE models. LLEP redistributes expert workloads across GPUs at runtime to eliminate straggler GPUs caused by token routing imbalance in models like DeepSeek-V3, Kimi-K2, and Qwen3.

**Paper:** [Least-Loaded Expert Parallelism: Load Balancing An Imbalanced Mixture-of-Experts](https://arxiv.org/abs/2601.17111)

## How LLEP Works

Standard EP assigns each expert to a fixed GPU. When routing is imbalanced, "hot" GPUs become bottlenecks while others idle. LLEP fixes this by:

1. **Gather** per-rank expert token counts via `all_gather`
2. **Plan** a Longest-Processing-Time (LPT) assignment that splits overloaded experts across helper GPUs
3. **Transfer** expert weights to helper GPUs via P2P
4. **Dispatch** tokens via modified AllToAll routing
5. **Compute** normal moe forward pass on each GPU's balanced workload
6. **Combine** results via inverse AllToAll back to original token order

## How to use it

Enable via TOML:

```toml
[llep]
enabled = true
min_tokens_per_gemm = 1
adaptive_threshold = 0.0
```

## Benchmark

**Model:** `debugmodel_ep8_llep_3b` — 1.75B total params, 64 experts, EP=8, top\_k=8, lbs=6, seq=4096
**Hardware:** 8×B200, 20 steps, no compile, no activation checkpointing

### Speed (steps 5–20 average)

| | With LLEP | Without LLEP | Delta |
|---|---|---|---|
| **Mean TPS** | 54,483 | 38,596 | **+41.2%** |

### Per-GPU Memory (step 5)

<details>
<summary><b>With LLEP</b> — all GPUs balanced within 1.86 GiB</summary>

| GPU | Memory | Memory % | TPS |
|-----|--------|----------|-----|
| 0 | 41.68 GiB | 23.4% | 55,698 |
| 1 | 40.97 GiB | 23.0% | 56,287 |
| 2 | 41.05 GiB | 23.0% | 54,551 |
| 3 | 40.53 GiB | 22.7% | 56,998 |
| 4 | 40.99 GiB | 23.0% | 56,409 |
| 5 | 42.39 GiB | 23.8% | 57,000 |
| 6 | 41.16 GiB | 23.1% | 56,797 |
| 7 | 41.28 GiB | 23.1% | 56,774 |
| **Spread** | **1.86 GiB** | | |

</details>

<details>
<summary><b>Without LLEP</b> — imbalanced, 13.11 GiB spread</summary>

| GPU | Memory | Memory % | TPS |
|-----|--------|----------|-----|
| 0 | 56.54 GiB | 31.7% | 38,697 |
| 1 | 52.40 GiB | 29.4% | 39,475 |
| 2 | 50.64 GiB | 28.4% | 39,465 |
| 3 | **57.48 GiB** | **32.2%** | 39,274 |
| 4 | 49.39 GiB | 27.7% | 39,534 |
| 5 | 44.37 GiB | 24.9% | 39,211 |
| 6 | 47.08 GiB | 26.4% | 39,480 |
| 7 | 55.08 GiB | 30.9% | 39,536 |
| **Spread** | **13.11 GiB** | | |

</details>

**Loss at step 20:** 3.081 (LLEP) vs 3.082 (no LLEP) — numerically identical.

### Reproduce

```bash
cd torchtitan

# WITH LLEP (20 steps)
torchrun --nproc_per_node=8 -m torchtitan.train \
  --job.config_file torchtitan/models/deepseek_v3/train_configs/debug_model_ep8_llep_3b.toml \
  --training.steps 20 --compile.no-enable

# WITHOUT LLEP (20 steps, same model)
torchrun --nproc_per_node=8 -m torchtitan.train \
  --job.config_file torchtitan/models/deepseek_v3/train_configs/debug_model_ep8_llep_3b.toml \
  --training.steps 20 --compile.no-enable --llep.enabled=False
```

## Test Plan

- [x] `test_llep.py` — 3/3 passed (LPT planning, CPU)
- [x] `test_llep_correctness.py` — 5/5 passed (Triton kernels, 1 GPU)
- [x] `test_llep_hooks.py` — 57/59 passed (hook-based flow, 8 GPUs)
- [x] `test_moe_ep_e2e.py` — 28/28 passed (full MoE forward+backward, 2 GPUs)
- [x] E2E training 20 steps WITH LLEP — loss decreasing, memory balanced
- [x] E2E training 20 steps WITHOUT LLEP — same loss, 7× higher memory spread

```bash
# Unit tests (no GPU required)
python -m pytest tests/unit_tests/test_llep.py -v
python -m pytest tests/unit_tests/test_llep_correctness.py -v

# Distributed tests (requires 2+ GPUs)
torchrun --nproc_per_node=2 tests/unit_tests/test_llep_hooks.py
torchrun --nproc_per_node=2 -m pytest tests/unit_tests/test_moe_ep_e2e.py -v
```

### CODE REVIEW

File: torchtitan/models/moe/moe.py [hunk 1]
```diff
@@ -30,6 +30,74 @@ def moe_init_std(dim_in: int, n_layers: int) -> float:
     return (2 / (dim_in * n_layers)) ** 0.5
 
 
+def fast_init_trunc_normal_(
```
Comment (jquesnelle) [inline torchtitan/models/moe/moe.py]: I think we're redefining these functions (see they're imported above from `utils`)

this pr added them -- we had independently fixed 2342 by doing the fast init -- let's just put the fast init into the `utils` version to minimize changes
https://github.com/pytorch/torchtitan/pull/2342

we should fold these changes into the utils version

**Review by jquesnelle — CHANGES_REQUESTED** [review]

**Comment by xrsrke** [comment]
@claude
